examples/src/main/java/kafka/examples/ExactlyOnceMessageProcessor.java

```diff
@@ -0,0 +1,190 @@
+                }
+
+                @Override
+                public void onPartitionsAssigned(Collection<TopicPartition> partitions) {
+                    printWithTxnId("Received partition assignment after rebalancing: " + partitions);
+                    messageRemaining.set(messagesRemaining(consumer));
+                }
+            });
+        } else {
+            // Do a range assignment of topic partitions.
+            List<TopicPartition> topicPartitions = new ArrayList<>();
+            int rangeSize = numPartitions / numInstances;
+            int startPartition = rangeSize * instanceIdx;
+            int endPartition = Math.min(numPartitions - 1, startPartition + rangeSize - 1);
+            for (int partition = startPartition; partition <= endPartition; partition++) {
+                topicPartitions.add(new TopicPartition(inputTopic, partition));
+            }
+
+            consumer.assign(topicPartitions);
+            printWithTxnId("Manually assign partitions: " + topicPartitions);
+        }
+
+        int messageProcessed = 0;
+        while (messageRemaining.get() > 0) {
+            ConsumerRecords<Integer, String> records = consumer.poll(Duration.ofMillis(200));
+            if (records.count() > 0) {
+                try {
+                    // Begin a new transaction session.
+                    producer.beginTransaction();
+                    for (ConsumerRecord<Integer, String> record : records) {
+                        // Process the record and send to downstream.
+                        ProducerRecord<Integer, String> customizedRecord = transform(record);
+                        producer.send(customizedRecord);
+                    }
+                    Map<TopicPartition, OffsetAndMetadata> positions = new HashMap<>();
+                    for (TopicPartition topicPartition : consumer.assignment()) {
+                        positions.put(topicPartition, new OffsetAndMetadata(consumer.position(topicPartition), null));
+                    }
+                    // Checkpoint the progress by sending offsets to group coordinator broker.
+                    // Under group mode, we must apply consumer group metadata for proper fencing.
+                    if (this.mode.equals("groupMode")) {
+                        producer.sendOffsetsToTransaction(positions, consumer.groupMetadata());
+                    } else {
+                        producer.sendOffsetsToTransaction(positions, consumerGroupId);
+                    }
+
+                    // Finish the transaction. All sent records should be visible for consumption now.
+                    producer.commitTransaction();
+                    messageProcessed += records.count();
+                } catch (CommitFailedException e) {
+                    // In case of a retriable exception, suggest aborting the ongoing transaction first for correctness.
+                    producer.abortTransaction();
```

Comment:
prop: abortTransaction can also throw ProducerFenced.

Reply:
Yea, which is ok as we will throw on ProducerFenced anyway

Reply:
Not sure I understand: we try to capture ProducerFenced below and wrap it as a KafkaException, but here if we throw ProducerFenced it would not be captured and wrapped, is that intentional?

Reply:
Oh I see your pt.